datafusion/src/sql/planner.rs

```diff
@@ -85,30 +86,40 @@ pub struct SqlToRel<'a, S: ContextProvider> {
     schema_provider: &'a S,
 }
 
-fn plan_key(key: Value) -> ScalarValue {
-    match key {
-        Value::Number(s, _) => ScalarValue::Int64(Some(s.parse().unwrap())),
-        Value::SingleQuotedString(s) => ScalarValue::Utf8(Some(s)),
-        _ => unreachable!(),
-    }
+fn plan_key(key: SQLExpr) -> Result<ScalarValue> {
+    let scalar = match key {
+        SQLExpr::Value(Value::Number(s, _)) => {
+            ScalarValue::Int64(Some(s.parse().unwrap()))
+        }
+        SQLExpr::Value(Value::SingleQuotedString(s)) => ScalarValue::Utf8(Some(s)),
+        _ => {
+            return Err(DataFusionError::SQL(ParserError(format!(
+                "Unsuported index key expression: {}",
+                key
+            ))))
+        }
+    };
+
+    Ok(scalar)
 }
 
-#[allow(clippy::branches_sharing_code)]
-fn plan_indexed(expr: Expr, mut keys: Vec<Value>) -> Expr {
-    if keys.len() == 1 {
-        let key = keys.pop().unwrap();
-        Expr::GetIndexedField {
-            expr: Box::new(expr),
-            key: plan_key(key),
-        }
+fn plan_indexed(expr: Expr, mut keys: Vec<SQLExpr>) -> Result<Expr> {
```

Comment:
I had to change this code anyways to handle arbitrary expressions due to https://github.com/sqlparser-rs/sqlparser-rs/pull/382 so I took the time to remove the clippy warning too